README.md
# Yarchi
### **Y**et **A**nother **R**ust **C**HIP-8 **I**nterpreter!

Yarchi is a simple yet functional CHIP-8 interpreter (emulator).

## Features
- Runs CHIP-8 programs and games (obviously!)
- Adjustable screen size
- Adjustable emulation speed
- Adjustable fore- and background color
- Possibility to run in CHIP-48 mode (required for some games to function properly)
- Debug mode

## Prerequisites
The interpreter depends on SDL2 library in order to handle video, audio and inputs. It must be therefore installed on your system.

On GNU/Linux operating systems this can be as simple as installing (example from Ubuntu 20.04):
```
apt install libsdl2-2.0-0
```

or on a MacOS (with `brew`):
```
brew install sdl2
```
For more information please see https://formulae.brew.sh/formula/sdl2.

## Running the interpreter
Simply run from your command line, while specifying the ROM path and passing arguments / flags.

For example, to run a program in debug mode with large screen:

```
./yarchi -d --screen-size large path/to/your/program.rom
```

## Building the interpreter
Yarchi depends only on a handful of dependencies and can be built very simply with the standard Rust toolchain by running from within the root directory:

```
cargo build
```

## Controls
Controls use the 'typical' mapping which is the following:

*(left: mapped on keyboard, right: original controls)*
| | | | | | | | | |
|-|-|-|-|-|-|-|-|-|
|1|2|3|4| |1|2|3|C|
|Q|W|E|R| |4|5|6|D|
|A|S|D|F| |7|8|9|E|
|Z|X|C|V| |A|B|C|D|

*the mapping is independent of your keyboard layout (i.e. bottom left key, whether `Z` or `Y`, is always `A`)

## Debug Mode
The interpreter also contains a debug mode, which can be 'activated' by passing `-d` or `--debug`. This mode makes it possible to inspect CHIP-8's state (program counter value, register values etc.), to pause / resume emulation and to execute emulation cycles one by one (when paused).

The described actions are available upon pressing:
- `P` - prints current state of CHIP-8
- `End` - pause/resume emulation
- `PgDown` - executes next cycle (4 instructions; possible only if emulation is paused)

## Contributions and collaboration
Something's not working right? Do you want to add a feature to the interpreter? Are you building one yourself and have questions?

Pull requests as well as opening an issue is more than welcome!

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0] - 2021-02-03
### Added
- Initial interpreter version

src/interpreter.rs
use std::thread::sleep;
use std::time::{Duration, Instant};

use sdl2::{audio::AudioQueue, audio::AudioSpecDesired, rect::Rect};
use sdl2::{event::Event, keyboard::Scancode, EventPump};
use sdl2::{render::Canvas, video::Window};

mod chip8;
pub mod config;

use crate::InterpErr;
use chip8::Chip8;
use config::Config;

// Keeping the main loop at the timer frequency;
// the instruction execution is then "synced" to
// this frequency, e.g. for 500 Hz interpreter frequency,
// this amounts to 500 / 60 ~= 8 cycles, or (assuming
// each opcode execution takes 2 cycles) 4 instructions.
const MAIN_LOOP_FREQUENCY: u32 = 60;

// How many microseconds to sleep, in order to sync at 60 Hz;
// the duration of actual code execution should be subtracted.
const SLEEP_TIME: u128 = ((100 / MAIN_LOOP_FREQUENCY) * 10000) as u128;

// Display size, i.e. how many 'points'.
pub const DISPLAY_WIDTH: usize = 64;
pub const DISPLAY_HEIGHT: usize = 32;

pub struct Interpreter {
    machine: Chip8,
    video: Canvas<Window>,
    audio: AudioQueue<i16>,
    events: EventPump,
    config: Config,
    debug: Debug,
}

struct Debug {
    running: bool,
    step_exec: bool,
}

impl Interpreter {
    pub fn new(
        sdl_ctx: &sdl2::Sdl,
        rom: Option<&str>,
        config: Config,
    ) -> Result<Interpreter, InterpErr> {
        let emu = Interpreter {
            machine: Chip8::new(config.c48_mode).load_program_to_memory(rom.unwrap())?,
            video: Interpreter::initiate_video(sdl_ctx, &config)?,
            audio: Interpreter::initiate_audio(sdl_ctx)?,
            events: sdl_ctx.event_pump()?,
            config,
            debug: Debug {
                running: true,
                step_exec: false,
            },
        };

        return Ok(emu);
    }


    pub fn run(&mut self) -> Result<(), InterpErr> {
        let mut previous_time: Instant;

        // run the main loop at 60 Hz
        'main_loop: loop {
            previous_time = Instant::now();

            let event_iter: Vec<Event> = self.events.poll_iter().collect();
            for event in event_iter {
                match event {
                    Event::Quit { .. } => break 'main_loop,
                    Event::KeyUp { scancode, .. } => {
                        if self.config.debug_mode {
                            self.handle_debug_input(scancode.unwrap());
                        }
                    }
                    _ => {}
                }
            }

            // if in debug mode & paused, skip execution
            if self.is_paused() {
                continue;
            }

            self.handle_timers();
            self.register_pressed_keys();

            for _ in 0..self.config.instructions_per_cycle() {
                self.machine.run_instruction(self.debug.step_exec)
            }

            self.refresh_screen()?;
            self.reset_exec_step();

            self.handle_loop_sync(Instant::now().duration_since(previous_time));

        }

        Ok(())
    }


    fn initiate_video(sdl_ctx: &sdl2::Sdl, config: &Config) -> Result<Canvas<Window>, InterpErr> {
        let video_subsys = sdl_ctx.video()?;

        let win = video_subsys
            .window(
                crate_name!(),
                64 * config.screen_size,
                32 * config.screen_size,
            )
            .position_centered()
            .build()?;

        let mut canvas = win.into_canvas().software().build()?;
        canvas.set_draw_color(config.background_color);
        canvas.clear();
        canvas.set_draw_color(config.foreground_color);

        canvas.present();
        Ok(canvas)
    }

    fn initiate_audio(sdl_ctx: &sdl2::Sdl) -> Result<AudioQueue<i16>, InterpErr> {
        let audio_subsys = sdl_ctx.audio()?;
        let audio_queue = audio_subsys.open_queue::<i16, _>(
            None,
            &AudioSpecDesired {
                freq: Some(44_100),
                channels: Some(1),
                samples: Some(4),
            },
        )?;
        audio_queue.queue(&generate_sound());
        Ok(audio_queue)
    }

    fn refresh_screen(&mut self) -> Result<(), InterpErr> {
        if !self.machine.screen.should_refresh() {
            return Ok(());
        }

        self.video.set_draw_color(self.config.background_color);
        self.video.clear();
        self.video.set_draw_color(self.config.foreground_color);

        for x in 0..64 {
            for y in 0..32 {
                let xy = (y * DISPLAY_WIDTH) + x;

                if self.machine.screen.display[xy] {
                    let r = Rect::new(
                        (x as u32 * self.config.screen_size) as i32,
                        (y as u32 * self.config.screen_size) as i32,
                        self.config.screen_size,
                        self.config.screen_size,
                    );
                    self.video.fill_rect(r)?;
                    self.video.draw_rect(r)?;
                }
            }
        }

        self.video.present();
        Ok(())
    }

    fn handle_timers(&mut self) {
        if self.machine.delay_timer > 0 {
            self.machine.delay_timer -= 1;
        }

        // as long as sound timer is > 0, emit beep
        if self.machine.sound_timer > 0 {
            self.machine.sound_timer -= 1;
            self.audio.queue(&generate_sound());
            self.audio.resume();
        } else {
            self.audio.pause();
        }
    }

    fn register_pressed_keys(&mut self) {
        self.machine.input.clear();
        let keyb_state = self.events.keyboard_state();
        for k in keyb_state.pressed_scancodes().into_iter() {
            match k {
                Scancode::Num1 => self.machine.input.push(0x1),
                Scancode::Num2 => self.machine.input.push(0x2),
                Scancode::Num3 => self.machine.input.push(0x3),
                Scancode::Num4 => self.machine.input.push(0xC),
                Scancode::Q => self.machine.input.push(0x4),
                Scancode::W => self.machine.input.push(0x5),
                Scancode::E => self.machine.input.push(0x6),
                Scancode::R => self.machine.input.push(0xD),
                Scancode::A => self.machine.input.push(0x7),
                Scancode::S => self.machine.input.push(0x8),
                Scancode::D => self.machine.input.push(0x9),
                Scancode::F => self.machine.input.push(0xE),
                Scancode::Y => self.machine.input.push(0xA),
                Scancode::X => self.machine.input.push(0x0),
                Scancode::C => self.machine.input.push(0xB),
                Scancode::V => self.machine.input.push(0xF),
                _ => {}
            }
        }
    }

    fn handle_debug_input(&mut self, scancode: Scancode) {
        match scancode {
            // Debug
            Scancode::P => println!("{:?}", self.machine),
            Scancode::End => {
                println!(
                    "Toggling interpreter state to - running: {}",
                    !self.debug.running
                );
                self.toggle_state()
            }
            Scancode::PageDown => {
                // ignore if interpreter not paused
                if self.debug.running {
                    return;
                }

                println!("Running next cycle");
                self.debug.step_exec = true;
            }
            _ => (),
        }
    }

    fn handle_loop_sync(&mut self, elapsed: Duration) {
        sleep(Duration::from_micros(
            SLEEP_TIME.checked_sub(elapsed.as_micros()).unwrap_or(0u128) as u64,
        ));
    }

    fn is_paused(&self) -> bool {
        self.config.debug_mode && !self.debug.running && !self.debug.step_exec
    }

    fn toggle_state(&mut self) {
        self.debug.running = !self.debug.running
    }

    fn reset_exec_step(&mut self) {
        if self.debug.step_exec {
            self.debug.step_exec = false
        }
    }
}

// generate a square wave
fn generate_sound() -> Vec<i16> {
    let tone_volume = 1_000i16;
    let period = 48_000 / 256;
    let mut result = Vec::new();

    for x in 0..8_000 {
        result.push(if (x / period) % 2 == 0 {
            tone_volume
        } else {
            -tone_volume
        });
    }
    result
}


src/interpreter/chip8.rs
use core::fmt;

use num_traits::int::PrimInt;

mod instruction;
mod operations;
mod program_counter;

use crate::interpreter::config::FONT;
use crate::interpreter::{DISPLAY_HEIGHT, DISPLAY_WIDTH};
use instruction::Instruction;
use operations::MAIN_TABLE as OP_TABLE;
use program_counter::ProgramCounter;

pub struct Chip8 {
    memory: [u8; 4096],
    pc: ProgramCounter,
    v: [u8; 16],
    i: u16,
    stack: Vec<u16>,
    pub screen: Screen,
    pub delay_timer: u8,
    pub sound_timer: u8,
    pub input: Vec<u8>,
    c48_mode: bool,
}

pub struct Screen {
    pub display: [bool; DISPLAY_WIDTH * DISPLAY_HEIGHT],
    pub refresh: bool,
}

enum Reg {
    V(usize),
    I,
}

impl Screen {
    fn new() -> Screen {
        return Screen{
            display: [false; DISPLAY_WIDTH * DISPLAY_HEIGHT],
            refresh: false,
        }
    }

    fn clear(&mut self) {
        self.display = [false; 2048];
    }

    pub fn should_refresh(&mut self) -> bool {
        if !self.refresh {
            return false;
        }

        self.refresh = false;
        true
    }
}

impl Chip8 {
    pub fn new(c48_mode: bool) -> Chip8 {
        return Chip8 {
            memory: [0; 4096],
            pc: ProgramCounter::new_with_value(0x200), // program starts at 0x200
            v: [0; 16],
            i: 0,
            stack: Vec::<u16>::new(),
            screen: Screen::new(),
            delay_timer: 0,
            sound_timer: 0,
            input: Vec::<u8>::new(),
            c48_mode,
        }
        .load_font();
    }

    pub fn load_program_to_memory(mut self, path: &str) -> Result<Chip8, std::io::Error> {
        let f = std::fs::read(path)?;
        let mut addr = 0x200; // program starts at 0x200

        for b in f {
            self.memory[addr] = b;
            addr += 1;
        }

        Ok(self)
    }

    fn load_font(mut self) -> Chip8 {
        let mut addr = 0x050;

        for &f in FONT.iter() {
            self.memory[addr] = f;
            addr += 1;
        }

        self
    }

    pub fn run_instruction(&mut self, is_debug: bool) {
        let instr = self.fetch();
        if is_debug {
            println!("Executed instr: Ox{:X}", instr.to_raw_instr())
        }

        self.decode_and_execute(instr)
    }

    fn decode_and_execute(&mut self, instr: Instruction) {
        OP_TABLE[instr.first_nibble() as usize](self, instr);
    }

    fn fetch(&mut self) -> Instruction {
        let addr = self.pc.value() as usize;

        // read 2 successive bytes from memory.
        let instr = Instruction::new_from_bytes(self.memory[addr], self.memory[addr + 1]);
        self.pc.increment();

        return instr;
    }

    fn set_reg_to<T: PrimInt>(&mut self, r: Reg, val: T) {
        match r {
            Reg::V(x) => {
                self.v[x] = val.to_u8().unwrap_or_default();
            }
            Reg::I => {
                self.i = val.to_u16().unwrap_or_default();
            }
        }
    }

    fn add_to_reg<T: PrimInt>(&mut self, r: Reg, val: T) {
        match r {
            Reg::V(x) => {
                self.v[x] = self.v[x].wrapping_add(val.to_u8().unwrap_or_default());
            }
            Reg::I => {
                self.i = self.i.wrapping_add(val.to_u16().unwrap_or_default());
            }
        }
    }

    fn set_register_flag_if_else_0(&mut self, condition: bool) {
        if condition {
            self.set_reg_to(Reg::V(15), 1u8)
        } else {
            self.set_reg_to(Reg::V(15), 0u8)
        }
    }

    fn draw(&mut self, instr: Instruction) {
        let (x, y) = instr.x_y();
        let mut display_x = (self.v[x] % 64) as usize;
        let mut display_y = (self.v[y] % 32) as usize;

        let bytes_to_read = instr.last_nibble();
        let addr = self.i as usize;
        self.set_reg_to(Reg::V(15), 0u8);

        for i in 0..bytes_to_read as usize {
            let mut sprite_byte = self.memory[addr + i];

            // iterate over all bits of current sprite byte.
            for _ in 0..8 {
                let current_pos = (display_y * DISPLAY_WIDTH) + display_x;

                // break if out of bounds.
                if current_pos >= 2048 {
                    break;
                }

                // if sprite bit is set, flip the display point;
                // if both are on, set flag register.
                if sprite_byte & 0x80 == 0x80 {
                    if self.screen.display[current_pos] {
                        self.screen.display[current_pos] = false;
                        self.set_reg_to(Reg::V(15), 1u8);
                    } else {
                        self.screen.display[current_pos] = true;
                    }
                }
                sprite_byte <<= 1;
                display_x += 1;
            }

            // reset the X position.
            display_x = display_x.checked_sub(8).unwrap_or(0);
            display_y += 1;
        }
    }

    fn handle_unknown_instr(&mut self, instr: Instruction) {
        println!(
            "Warning: Unknown instruction 0x{:X} at program counter {}; skipping",
            instr.to_raw_instr(),
            self.pc.value()
        )
    }
}

impl std::fmt::Debug for Chip8 {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut ds = f.debug_struct("CHIP-8 State");
        ds.field("Program Counter: ", &self.pc.value());

        for i in 0..16 {
            let name = format!("Register V[{}]", i);
            ds.field(name.as_str(), &self.v[i]);
        }

        ds.field("Register I", &self.i);
        ds.field("Delay Time", &self.delay_timer);
        ds.field("Sound Time", &self.sound_timer);
        ds.finish()
    }
}

#[cfg(test)]
mod tests {
    use super::Chip8;
    use super::Instruction;
    #[test]
    fn set_register_instructions_are_decoded_and_executed() {
        let mut machine = Chip8::new(false);

        machine.decode_and_execute(Instruction::new_from_bytes(0x65, 0x42));
        assert_eq!(machine.v[5], 0x42);

        machine.decode_and_execute(Instruction::new_from_bytes(0x86, 0x50));
        assert_eq!(machine.v[6], 0x42);
    }

    #[test]
    fn pressed_key_instruction_is_decoded_and_executed() {
        let mut machine = Chip8::new(false);

        machine.decode_and_execute(Instruction::new_from_bytes(0x65, 0xA));
        assert_eq!(machine.v[5], 0xA);

        machine.input.push(0xA);
        machine.decode_and_execute(Instruction::new_from_bytes(0xE5, 0x9E));
        assert_eq!(machine.pc.value(), 0x202);
    }

    #[test]
    fn b_c_d_instruction_is_decoded_and_executed() {
        let mut machine = Chip8::new(false);

        machine.decode_and_execute(Instruction::new_from_bytes(0x61, 0x7B));
        assert_eq!(machine.v[1], 0x7B);
        machine.decode_and_execute(Instruction::new_from_bytes(0xA6, 0x66));
        assert_eq!(machine.i, 0x666);
        machine.decode_and_execute(Instruction::new_from_bytes(0xF1, 0x33));

        assert_eq!(machine.memory[0x666], 1);
        assert_eq!(machine.memory[0x667], 2);
        assert_eq!(machine.memory[0x668], 3);
    }

    #[test]
    fn draw_instructions_are_decoded_and_executed() {
        let mut machine = Chip8::new(false);

        machine.decode_and_execute(Instruction::new_from_bytes(0x60, 0x1));
        assert_eq!(machine.v[0], 0x1);
        machine.decode_and_execute(Instruction::new_from_bytes(0x61, 0x1));
        assert_eq!(machine.v[1], 0x1);

        machine.decode_and_execute(Instruction::new_from_bytes(0xA6, 0x66));
        assert_eq!(machine.i, 0x666);
        machine.memory[0x666] = 0b01010101;

        machine.decode_and_execute(Instruction::new_from_bytes(0xD0, 0x11));
        assert_eq!(machine.screen.display[65], false);
        assert_eq!(machine.screen.display[66], true);
        assert_eq!(machine.screen.display[67], false);
        assert_eq!(machine.screen.display[68], true);
        assert_eq!(machine.screen.display[69], false);
        assert_eq!(machine.screen.display[70], true);
        assert_eq!(machine.screen.display[71], false);
        assert_eq!(machine.screen.display[72], true);
    }
}


src/interpreter/config.rs
use sdl2::pixels::Color;

use crate::interpreter::MAIN_LOOP_FREQUENCY;

// Used to calculate the actual screen size from configuration.
const DEFAULT_SCREEN_SIZE_COEFF: u32 = 10;

// Default frequency to use.
const DEFAULT_EMU_FREQUENCY: u32 = 500;

const DEFAULT_BACKGROUND_COLOR: Color = Color::RGB(0, 0, 0);
const DEFAULT_FOREGROUND_COLOR: Color = Color::RGB(0, 255, 102);

// 16 characters, each represented by 5 bytes
pub const FONT: [u8; 0x10 * 5] = [
    0xF0, 0x90, 0x90, 0x90, 0xF0, // 0
    0x20, 0x60, 0x20, 0x20, 0x70, // 1
    0xF0, 0x10, 0xF0, 0x80, 0xF0, // 2
    0xF0, 0x10, 0xF0, 0x10, 0xF0, // 3
    0x90, 0x90, 0xF0, 0x10, 0x10, // 4
    0xF0, 0x80, 0xF0, 0x10, 0xF0, // 5
    0xF0, 0x80, 0xF0, 0x90, 0xF0, // 6
    0xF0, 0x10, 0x20, 0x40, 0x40, // 7
    0xF0, 0x90, 0xF0, 0x90, 0xF0, // 8
    0xF0, 0x90, 0xF0, 0x10, 0xF0, // 9
    0xF0, 0x90, 0xF0, 0x90, 0x90, // A
    0xE0, 0x90, 0xE0, 0x90, 0xE0, // B
    0xF0, 0x80, 0x80, 0x80, 0xF0, // C
    0xE0, 0x90, 0x90, 0x90, 0xE0, // D
    0xF0, 0x80, 0xF0, 0x80, 0xF0, // E
    0xF0, 0x80, 0xF0, 0x80, 0x80, // F
];

pub struct Config {
    pub screen_size: u32,
    emu_speed: u32,
    pub background_color: Color,
    pub foreground_color: Color,
    pub debug_mode: bool,
    pub c48_mode: bool,
}

impl Config {
    pub fn from_args(matches: &clap::ArgMatches<'_>) -> Config {
        if matches.is_present("debug") {
            println!("{}\n{}", "Entering debug mode...", crate::DEBUG_MSG)
        }

        Config {
            screen_size: Config::set_screen_size(matches),
            emu_speed: Config::set_emu_frequency(matches),
            background_color: Config::set_color(matches, "bg-color")
                .unwrap_or(DEFAULT_BACKGROUND_COLOR),
            foreground_color: Config::set_color(matches, "fg-color")
                .unwrap_or(DEFAULT_FOREGROUND_COLOR),
            debug_mode: matches.is_present("debug"),
            c48_mode: matches.is_present("c48"),
        }
    }

    // Assuming each instruction takes 2 cycles.
    pub fn instructions_per_cycle(&self) -> u32 {
        return (self.emu_speed / MAIN_LOOP_FREQUENCY) / 2;
    }

    fn set_color(m: &clap::ArgMatches<'_>, arg: &str) -> Option<Color> {
        match m.value_of(arg) {
            Some(v) => {
                let rgb: Vec<&str> = v.split(",").collect();
                Some(Color::RGB(
                    rgb[0].parse::<u8>().unwrap(),
                    rgb[1].parse::<u8>().unwrap(),
                    rgb[2].parse::<u8>().unwrap(),
                ))
            }
            _ => None,
        }
    }

    fn set_screen_size(m: &clap::ArgMatches<'_>) -> u32 {
        match m.value_of("screen-size") {
            Some(v) => match v {
                "medium" => 12,
                "large" => 16,
                _ => DEFAULT_SCREEN_SIZE_COEFF,
            },
            _ => DEFAULT_SCREEN_SIZE_COEFF,
        }
    }

    fn set_emu_frequency(m: &clap::ArgMatches<'_>) -> u32 {
        match m.value_of("freq") {
            Some(v) => v.parse::<u32>().unwrap(),
            _ => DEFAULT_EMU_FREQUENCY,
        }
    }
}


src/interpreter/chip8/operations.rs
use super::instruction::Instruction;
use super::{Chip8, Reg};

use rand::prelude::*;

pub const MAIN_TABLE: [fn(&mut Chip8, Instruction); 16] = [
    op_table_0, op_1nnn, op_2nnn, op_3xkk, op_4xkk, op_5xy0, op_6xkk, op_7xkk, op_table_8, op_9xy0,
    op_annn, op_bnnn, op_cxkk, op_dxyn, op_table_e, op_table_f,
];

fn op_table_0(c: &mut Chip8, instr: Instruction) {
    match instr.kk() {
        0xE0 => {
            c.screen.clear();
            c.screen.refresh = true;
        }
        0xEE => c.pc.set_to(c.stack.pop().unwrap_or_default()),
        _ => c.handle_unknown_instr(instr),
    }
}

fn op_table_8(c: &mut Chip8, instr: Instruction) {
    let (x, y) = instr.x_y();
    match instr.last_nibble() {
        0x0 => c.set_reg_to(Reg::V(x), c.v[y]),
        0x1 => c.set_reg_to(Reg::V(x), c.v[x] | c.v[y]),
        0x2 => c.set_reg_to(Reg::V(x), c.v[x] & c.v[y]),
        0x3 => c.set_reg_to(Reg::V(x), c.v[x] ^ c.v[y]),
        0x4 => {
            let (val, overflow) = c.v[x].overflowing_add(c.v[y]);
            c.set_register_flag_if_else_0(overflow);
            c.set_reg_to(Reg::V(x), val);
        }
        0x5 => {
            let (val, overflow) = c.v[x].overflowing_sub(c.v[y]);
            c.set_register_flag_if_else_0(!overflow);
            c.set_reg_to(Reg::V(x), val);
        }
        0x6 => {
            if !c.c48_mode {
                c.set_reg_to(Reg::V(x), c.v[y])
            };
            let (val, overflow) = c.v[x].overflowing_shr(1);
            c.set_register_flag_if_else_0(overflow);
            c.set_reg_to(Reg::V(x), val);
        }
        0x7 => {
            let (val, overflow) = c.v[y].overflowing_sub(c.v[x]);
            c.set_register_flag_if_else_0(!overflow);
            c.set_reg_to(Reg::V(x), val);
        }
        0xE => {
            if !c.c48_mode {
                c.set_reg_to(Reg::V(x), c.v[y])
            };
            let (val, overflow) = c.v[x].overflowing_shl(1);
            c.set_register_flag_if_else_0(!overflow);
            c.set_reg_to(Reg::V(x), val);
        }
        _ => c.handle_unknown_instr(instr),
    }
}

fn op_table_e(c: &mut Chip8, instr: Instruction) {
    match instr.kk() {
        0x9E => c.pc.increment_if(c.input.contains(&c.v[instr.x()])),
        0xA1 => c.pc.increment_if(!c.input.contains(&c.v[instr.x()])),
        _ => c.handle_unknown_instr(instr),
    }
}
fn op_table_f(c: &mut Chip8, instr: Instruction) {
    let x = instr.x();
    match instr.kk() {
        0x07 => c.set_reg_to(Reg::V(x), c.delay_timer),
        0x0A => {
            c.pc.decrement_if((&c.input).is_empty());
            let key = c.input.first().unwrap_or(&0u8).to_owned();
            c.set_reg_to(Reg::V(x), key)
        }
        0x15 => c.delay_timer = c.v[x],
        0x18 => c.sound_timer = c.v[x],
        0x1E => {
            c.add_to_reg(Reg::I, c.v[x]);

            if c.i > 0x1000 {
                c.set_reg_to(Reg::V(15), 1u8);
            }
        }
        0x29 => {
            // get last nibble only for char.
            let ch = (c.v[x] & 0xF) as u16;
            // start address + offset to given character
            c.set_reg_to(Reg::I, 0x050 + (5 * ch));
        }
        0x33 => {
            let val = c.v[x];
            c.memory[c.i as usize] = val / 100;
            c.memory[(c.i + 1) as usize] = (val % 100) / 10;
            c.memory[(c.i + 2) as usize] = (val % 100) % 10;
        }
        0x55 => {
            for n in 0..x + 1 {
                c.memory[(c.i as usize + n) as usize] = c.v[n];
            }
        }
        0x65 => {
            for n in 0..x + 1 {
                c.set_reg_to(Reg::V(n), c.memory[(c.i as usize + n) as usize]);
            }
        }
        _ => c.handle_unknown_instr(instr),
    }
}

fn op_1nnn(c: &mut Chip8, instr: Instruction) {
    c.pc.set_to(instr.nnn())
}

fn op_2nnn(c: &mut Chip8, instr: Instruction) {
    c.stack.push(c.pc.value());
    c.pc.set_to(instr.nnn());
}

fn op_3xkk(c: &mut Chip8, instr: Instruction) {
    c.pc.increment_if(c.v[instr.x()] == instr.kk());
}

fn op_4xkk(c: &mut Chip8, instr: Instruction) {
    c.pc.increment_if(c.v[instr.x()] != instr.kk());
}

fn op_5xy0(c: &mut Chip8, instr: Instruction) {
    let (x, y) = instr.x_y();
    c.pc.increment_if(c.v[x] == c.v[y]);
}

fn op_6xkk(c: &mut Chip8, instr: Instruction) {
    c.set_reg_to(Reg::V(instr.x()), instr.kk());
}
fn op_7xkk(c: &mut Chip8, instr: Instruction) {
    c.add_to_reg(Reg::V(instr.x()), instr.kk());
}
fn op_9xy0(c: &mut Chip8, instr: Instruction) {
    let (x, y) = instr.x_y();
    c.pc.increment_if(c.v[x] != c.v[y]);
}
fn op_annn(c: &mut Chip8, instr: Instruction) {
    c.set_reg_to(Reg::I, instr.nnn());
}

// ambiguous OP: either BNNN or BXKK
fn op_bnnn(c: &mut Chip8, instr: Instruction) {
    if c.c48_mode {
        c.pc.set_to((instr.kk() + c.v[instr.x()]) as u16)
    } else {
        c.pc.set_to(instr.nnn() + c.v[0] as u16)
    }
}

fn op_cxkk(c: &mut Chip8, instr: Instruction) {
    let r: u8 = thread_rng().gen();
    c.set_reg_to(Reg::V(instr.x()), r & instr.kk());
}

fn op_dxyn(c: &mut Chip8, instr: Instruction) {
    c.draw(instr);
    c.screen.refresh = true;
}


src/interpreter/chip8/instruction.rs
#[derive(Copy, Clone)]
pub struct Instruction(u16);

impl Instruction {
    pub fn new_from_bytes(b1: u8, b2: u8) -> Instruction {
        return Instruction(u16::from_be_bytes([b1, b2]));
    }

    pub fn to_raw_instr(self) -> u16 {
        return self.0;
    }

    pub fn first_nibble(self) -> u8 {
        return ((self.0 & 0xF000) >> 12) as u8;
    }

    pub fn last_nibble(self) -> u8 {
        return (self.0 & 0xF) as u8;
    }

    pub fn x(self) -> usize {
        return ((self.0 & 0x0F00) >> 8) as usize;
    }

    pub fn x_y(self) -> (usize, usize) {
        return (
            ((self.0 & 0x0F00) >> 8) as usize,
            ((self.0 & 0x0F0) >> 4) as usize,
        );
    }

    pub fn nnn(self) -> u16 {
        return self.0 & 0x0FFF;
    }

    pub fn kk(self) -> u8 {
        return (self.0 & 0x00FF) as u8;
    }
}

#[cfg(test)]
mod tests {
    use super::Instruction;

    #[test]
    fn instruction_is_constructed_correctly() {
        let instr = Instruction::new_from_bytes(0x12, 0x34);
        assert_eq!(instr.to_raw_instr(), 0x1234);
    }

    #[test]
    fn instruction_is_read_correctly() {
        let instr = Instruction::new_from_bytes(0x12, 0x34);

        assert_eq!(instr.first_nibble(), 0x1);
        assert_eq!(instr.last_nibble(), 0x4);
        assert_eq!(instr.x(), 0x2);
        assert_eq!(instr.x_y(), (0x2, 0x3));
        assert_eq!(instr.nnn(), 0x234);
        assert_eq!(instr.kk(), 0x34);
    }
}


src/interpreter/chip8/program_counter.rs
#[derive(Copy, Clone)]
pub struct ProgramCounter(u16);

impl ProgramCounter {
    pub fn new_with_value(val: u16) -> ProgramCounter {
        ProgramCounter(val)
    }

    pub fn set_to(&mut self, val: u16) -> () {
        self.0 = val
    }

    pub fn value(&self) -> u16 {
        self.0
    }

    pub fn increment(&mut self) -> () {
        self.0 += 2
    }

    pub fn increment_if(&mut self, condition: bool) -> () {
        if condition {
            self.increment()
        }
    }

    pub fn decrement_if(&mut self, condition: bool) -> () {
        if condition {
            self.0 -= 2
        }
    }
}

#[cfg(test)]
mod tests {
    use super::ProgramCounter;

    #[test]
    fn program_counter_operates_correctly() {
        let mut pc = ProgramCounter::new_with_value(10);

        assert_eq!(pc.value(), 10);

        pc.set_to(11);
        assert_eq!(pc.value(), 11);

        pc.increment();
        assert_eq!(pc.value(), 13);

        pc.increment_if(true);
        assert_eq!(pc.value(), 15);

        pc.decrement_if(false);
        assert_eq!(pc.value(), 15);
    }
}


src/main.rs
#![warn(clippy::all, rust_2018_idioms)]

#[macro_use]
extern crate clap;

use crate::interpreter::config::Config;
use crate::interpreter::Interpreter;

use clap::{App, Arg, crate_authors, crate_description};

mod interpreter;

const DEBUG_MSG: &str = "
Available actions upon pressing:
- P - prints current state of CHIP-8
- End - pause/resume emulation
- PgDown - executes next cycle (4 instructions; possible only if emulation is paused)
";

type InterpErr = Box<dyn std::error::Error>;
type InterpResult<T> = Result<T, InterpErr>;

fn main() -> InterpResult<()> {
    let long_debug_msg = format!(
        "{}\n{}",
        "Enables debug mode, which allows for pausing emulation and executing cycles step-by-step.",
        DEBUG_MSG
    );

    let about_with_controls = format!(
        "{}\n{}",
        crate_description!(),
        "
To control the interpreter, use the left side of your keyboard.
The controls are mapped to the following keys:

|1|2|3|4|
|Q|W|E|R|
|A|S|D|F|
|Z|X|C|V|
",
    );

    let app = App::new(crate_name!())
        .version(crate_version!())
        .about(about_with_controls.as_str())
        .author(crate_authors!())
        .arg(
            Arg::with_name("screen-size")
                .takes_value(true)
                .long("screen-size")
                .short("s")
                .help("Sets the screen size to small (640x320, default), medium (768x384) or large (1024x512)")
                .possible_values(&["small", "medium", "large"])
        )
        .arg(
            Arg::with_name("freq")
                .takes_value(true)
                .long("interpreter-frequency")
                .short("f")
                .help("Adjusts emulation speed to specified Hz value. Valid values: 200-1000 Hz. Default: 500 Hz.")
                .validator(is_valid_emu_frequency)
        )
        .arg(
            Arg::with_name("debug")
                .long("debug")
                .short("d")
                .help("Enables debug mode, which allows for pausing emulation and executing cycles step-by-step.
For all available commands, print information with --help.")
                .long_help(long_debug_msg.as_str())
        )
        .arg(
            Arg::with_name("c48")
                .long("chip-48-mode")
                .short("c")
                .help("Executes certain instructions in a mode compatible with CHIP-48. Required for some programs.")
        )
        .arg(
            Arg::with_name("fg-color")
            .takes_value(true)
            .long("foreground-color")
            .help("Changes foreground color to specified RGB value. Format: R,G,B")
            .validator(is_valid_rgb_color)
        )
        .arg(
            Arg::with_name("bg-color")
            .takes_value(true)
            .long("background-color")
            .help("Changes background color to specified RGB value. Format: R,G,B")
            .validator(is_valid_rgb_color)
        )
        .arg(Arg::with_name("INPUT").required(true).help("Path to ROM which should be run"));

    let matches = app.get_matches();

    let sdl_ctx = sdl2::init()?;
    let mut interpreter = Interpreter::new(
        &sdl_ctx,
        matches.value_of("INPUT"),
        Config::from_args(&matches),
    )?;
    interpreter.run()?;

    Ok(())
}

fn is_valid_emu_frequency(freq: String) -> Result<(), String> {
    match freq.parse::<u16>() {
        Ok(f) => {
            if f < 200 || f > 1000 {
                return Err(
                    "invalid interpreter frequency specified: must be in range 200 - 1000 Hz"
                        .to_string(),
                );
            }

            Ok(())
        }
        Err(e) => Err(format!("parsing interpreter frequency failed: {}", e)),
    }
}

fn is_valid_rgb_color(rgb: String) -> Result<(), String> {
    let vals: Vec<&str> = rgb.split(",").collect();
    if vals.len() != 3 {
        return Err("invalid RGB value provided. Valid format: R,G,B".to_string());
    }

    for v in vals {
        match v.parse::<u8>() {
            Err(e) => return Err(format!("parsing RGB values failed: {}", e)),
            _ => return Ok(()),
        }
    }

    Ok(())
}
